### ISSUE TITLE
coqchk -o incorrectly reports abstracted subproofs from the obligation tactic in Program Definition as axioms in master

### ISSUE THREAD

**Issue by JasonGross** [coq/coq#13324]
#### Description of the problem

```
$ cat abs.v
```
```coq
Require Import Coq.Program.Basics Coq.Program.Tactics.
Obligation Tactic := abstract exact I.
Program Definition foo : True := _.
Print Assumptions foo.
```
```
$ for ver in master 8.12.0; do (true | ~/.local64/coq/coq-$ver/bin/coqtop -q); ~/.local64/coq/coq-$ver/bin/coqc -q abs.v; ~/.local64/coq/coq-$ver/bin/coqchk -silent -o abs.vo; done
Welcome to Coq jgross-Leopard-WS:/home/jgross/Downloads/coq/coq-master,master (cdfe69d6da6b32338ba74c9f599c74389089c9dd)
Skipping rcfile loading.

Coq <
Closed under the global context

CONTEXT SUMMARY
===============

* Theory: Set is predicative

* Axioms:
    abs.temporary_proof1_subproof

* Constants/Inductives relying on type-in-type: <none>

* Constants/Inductives relying on unsafe (co)fixpoints: <none>

* Inductives whose positivity is assumed: <none>

Welcome to Coq 8.12.0 (August 2020)
Skipping rcfile loading.

Coq <
Closed under the global context

CONTEXT SUMMARY
===============

* Theory: Set is predicative

* Axioms: <none>

* Constants/Inductives relying on type-in-type: <none>

* Constants/Inductives relying on unsafe (co)fixpoints: <none>

* Inductives whose positivity is assumed: <none>

```

This is a regression over 8.12.0.  Interestingly, it only occurs when the `abstract` is in the obligation tactic.

#### Coq Version

cdfe69d6da6b32338ba74c9f599c74389089c9dd

**Comment by ejgallego** [comment]
Thanks for reporting @JasonGross , bisect done with:
```
dune build theories/Program/{Basics,Tactics}.vo checker/coqchk.exe && dune exec -- dev/shim/coqc-prelude k.v && _build/default/checker/coqchk.exe -silent -o k.vo
```
faulty commit:
```
427e8cac2c267a7302b71979c0f1019598176e31 is the first bad commit
commit 427e8cac2c267a7302b71979c0f1019598176e31
Author: Pierre-Marie Pédrot <pierre-marie.pedrot@inria.fr>
Date:   Fri Aug 28 17:37:24 2020 +0200

    Drop opaque bodies of abstracted definitions.
    
    This should save us a lot of useless hashconsing. This change should not be
    observable because outside of the proof, the abstracted definition will be
    either inlined or redefined with the body coming from the side-effect.

 kernel/safe_typing.ml | 14 ++++++++------
 1 file changed, 8 insertions(+), 6 deletions(-)
```
introduced in #12935 , cc @ppedrot @SkySkimmer

**Comment by SkySkimmer** [comment]
Smaller example:
~~~coq
Definition foo : True := ltac:(abstract exact I).
~~~

**Comment by ppedrot** [comment]
Indeed the vo file does not contain the body of the abstracted proof...

**Comment by RalfJung** [comment]
We're also seeing this in std++. Would be nice to get this resolved so that we can again reliably check for axioms. :)

**Comment by JasonGross** [comment]
This shows up in a bunch of zify instances (not a complete list):
```
      Coq.micromega.ZifyInst.ltac_gen_subproof22
      Coq.micromega.ZifyInst.ltac_gen_subproof21
      Coq.micromega.ZifyInst.ltac_gen_subproof20
      Coq.micromega.ZifyInst.ltac_gen_subproof19
      Coq.micromega.ZifyInst.ltac_gen_subproof18
      Coq.micromega.ZifyInst.ltac_gen_subproof17
      Coq.micromega.ZifyInst.ltac_gen_subproof16
      Coq.micromega.ZifyInst.ltac_gen_subproof15
      Coq.micromega.ZifyInst.ltac_gen_subproof14
      Coq.micromega.ZifyInst.ltac_gen_subproof13
      Coq.micromega.ZifyInst.ltac_gen_subproof12
      Coq.micromega.ZifyInst.ltac_gen_subproof11
      Coq.micromega.ZifyInst.ltac_gen_subproof10
      Coq.micromega.ZifyInst.ltac_gen_subproof
```

**Comment by ppedrot** [comment]
I had another look at this issue and I don't think that 427e8cac2c267a7302b71979c0f1019598176e31 is the real cause of the problem, rather it exposes an issue with abstracted subproofs.

Actually, if you look at it the body of the proof is axiom-free, so this part is fine.
```coq
Definition foo : True := ltac:(abstract exact I).

Print foo.
(* foo = (fun ltac_gen_subproof : True => ltac_gen_subproof) I
     : True *)
```

The problem is that the global environment is incorrectly extended with a side-effect that lingers around although it should have disappeared right after the tactic-in-term returns.
```coq
Print ltac_gen_subproof.
(* *** [ ltac_gen_subproof : True ] *)
```
Before the above commit, it was less visible because the proof was axiom free, but in practice it did mean that we had a useless term in the vo file taking space for no good reason.

I suspect that wrapping the tactic-in-term mechanism in a purifying handler is probably enough to solve this but I am vaguely afraid that some eldritch parts of fiat-crypto-like developments rely on this *feature*.

**Comment by JasonGross** [comment]
I don't think we rely on abstracted subproofs from tactics in terms sticking around in fiat-crypto.  Instead we do something like `simple refine (let v : T := _ in _); [ abstract tac | ]; let val := (eval cbv delta [v] in v) in clear v; ...` to get access to abstracted names in a way that will have them stick around.

**Comment by ppedrot** [comment]
I distinctly remember seeing some calls to `abstract tac as foo` in fiat-crypto, or maybe at least fiat-crypto-legacy. This is a notorious issue with abstract because it ultimately prevents to have a fully purely functional implementation (name resolution depends on the global nametab). If it's actually not needed, then it's a good thing because it simplifies a lot one of the PRs I started writing some time ago.

**Comment by JasonGross** [comment]
Note that this is not just an issue with `coqchk`, but `Print Assumptions` and `Print` and `Extraction` all also believe that the subproof has no body.

**Comment by JasonGross** [comment]
> I distinctly remember seeing some calls to `abstract tac as foo` in fiat-crypto, or maybe at least fiat-crypto-legacy. This is a notorious issue with abstract because it ultimately prevents to have a fully purely functional implementation (name resolution depends on the global nametab). If it's actually not needed, then it's a good thing because it simplifies a lot one of the PRs I started writing some time ago.

Coming back to this belatedly: we do use `transparent abstract foo as bar`, but, IIRC, don't rely on the resulting constant having exactly the name `bar`; I believe instead we use tactic code to get the name of the constant, and the `as` is there just for user-readability of names.  It's possible in fiat-crypto-legacy we do in fact rely on the names, though (I'd have to check to be sure).

**Comment by ppedrot** [comment]
@JasonGross would you be fine with deprecating `abstract using` then? It's a major PITA w.r.t. functionalization of the abstract feature because it requires modifying the nametab imperatively.

**Comment by JasonGross** [comment]
It's a bit unfortunate (and doesn't metacoq need the same feature for `run_template (tmDefinition ...)` or whatever anyway?), but, yes, I'd be okay dropping our uses of `abstract using` if need be.  Better would be having a deprecation of the current behavior and a semantic change such that `abstract using NAME` always succeeds and gives a name that contains `NAME` somewhere inside it, but no guarantee is made about the exact name used.  (This alternative I'd be quite happy with.)

**Comment by JasonGross** [comment]
What's the requirement / design around functionalizing abstract?

**Comment by ppedrot** [comment]
It shouldn't create a constant while the proof is open, or at the very least not make it available through a user-facing name.

Basically, it's easy to replace with a special kind of evar (or something similar) that is handled in a special way at proof closure time. What is very hard to make work properly in the current design is the desynchronization between the nametab (the mapping name ↦ constant) and the proof state. I had a simple patch replacing the constant registering through the `using`  clause with a simple local variable definition of the same name, which was kind of working (up to printing issues), see #14937.

For transparent abstract, things are a bit different since we do want to have access to the body of the proof while not polluting the goal with it. I suspect we should implement abstract directly at the level of the proof engine, which would solve a lot of problems (e.g. the fact that you must solve the abstract goal with a single tactic invocation) and may subsume the obligation mechanism.

**Comment by JasonGross** [comment]
I think the usage pattern I have is generally
```coq
let H := fresh in
simple refine (let H := _ in _);
[ shelve | abstract tac | ];
let H := lazymatch goal with H := _ |- _ => H end in
let name := (eval cbv delta [H] in H) in
clear H;
do stuff with name
```
If `transparent abstract` is used instead of `abstract` then `eval cbv delta [name] in name` should get the body.  Ideally this should not pollute the context with a large type that other tactics see when walking the context, `clear` should not remove the name, etc.  Is this feasible with your proposal?  How does the local variable definition show up in goals other than the one `abstract` is invoked in?

**Comment by JasonGross** [comment]
> I suspect we should implement abstract directly at the level of the proof engine, which would solve a lot of problems (e.g. the fact that you must solve the abstract goal with a single tactic invocation) and may subsume the obligation mechanism.

This sounds good to me.  Having a "abstract the rest of this subgoal (transparently or otherwise)" sounds good to me, though it would have to be more powerful than the obligation mechanism currently is (obligations must be solved in dependency order).  I suspect it also wouldn't quite subsume obligations, which IIRC allow interleaved definitions without nested proof mode set.

**Comment by ppedrot** [comment]
As of adaf0357f40110d7c2d9badfd8a220d8c631a6d1, this is fixed. This bug was most probably fixed by #21284.
